lib/configure.js

```diff
@@ -334,6 +341,22 @@ module.exports = function() {
       if (err.message === 'repository not found') {
         process.stdout.write('I didn\'t find your repository in Travis.  Syncing Travis with GitHub…\n');
         return travis.users.sync.post()
+        .catch(function(err) {
+          // Ignore the exception if Travis is already syncing. The status code
+          // of the error is 409 and its message is 'Sync already in progress.
+          // Try again later.', but instead of relying on that we simply perform
+          // an explicit check.
+          return travisIsSyncing().then(function(is_syncing) {
+            if (!is_syncing) {
+              // Check again if it's active on Travis (in case it finished syncing
+              // right before we checked)
+              return ensureActiveInTravis().catch(function() {
+                // Throw the original error.
+                throw err;
+              });
+            }
+          });
+        })
         .then(travisAwaitSyncing)
```

Comment:
Both the *catch* function and *travisAwaitSyncing* call *travisIsSyncing*, which means that *travisIsSyncing* gets called twice in a row when Travis is already syncing, first by the *catch* function and then by *travisAwaitSyncing*. That isn't the end of the world, but it does result in an extra, unnecessary network request.

Another issue is that *ensureActiveInTravis* will get called twice if Travis finishes syncing right before the *catch* function calls *travisIsSyncing*: first by the *catch* function, and then by the *ensureActiveInTravis* link at the end of the chain.

To resolve these problems, I would make *travisAwaitSyncing* wait unconditionally and then recurse if Travis is still syncing:

```js
    function travisAwaitSyncing() {
      process.stdout.write('Waiting for Travis to finish syncing…\n');
      return new Promise(function(resolve, reject) { setTimeout(resolve, 5000) })
      .then(travisIsSyncing)
      .then(function(isSyncing) {
        if (isSyncing) {
          return travisAwaitSyncing();
        }
      });
    }
```

Then I would make the *catch* function explicitly ignore the `Sync already in progress…` error, so the promise chain resolution falls through to the existing *travisAwaitSyncing* call in the chain:

```js
    return ensureActiveInTravis()
    .catch(function(err) {
      if (err.message === 'repository not found') {
        process.stdout.write('I didn\'t find your repository in Travis.  Syncing Travis with GitHub…\n');
        return travis.users.sync.post()
        .catch(function(err) {
          if (/* status code is 409 && message is 'Sync already in progress… */) {
            // Ignore the exception if Travis is already syncing.
          } else {
            throw err;
          }
        })
        .then(travisAwaitSyncing)
        .then(ensureActiveInTravis);
      }
      throw err;
    });
```

As a side-effect, this also avoids calling *travisIsSyncing* immediately after requesting a sync (or finding out that Travis is already syncing), which seems worthwhile, since it's highly likely that Travis will still be syncing then.